src/Compatibility/Microsoft.DotNet.ApiCompat.Shared/RegexStringTransformer.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Text.RegularExpressions;
+
+namespace Microsoft.DotNet.ApiCompat
+{
+    /// <summary>
+    /// A regex string transformer that transforms an input string via a set of regex patterns and replacement strings.
+    /// </summary>
+    internal class RegexStringTransformer
+    {
+        private readonly (Regex Regex, string ReplacementString)[] _patterns;
+
+        /// <summary>
+        /// Initializes the regex string transformer with a given capture group and replacement patterns.
+        /// </summary>
+        /// <param name="captureGroupPattern">The capture group pattern to retrieve data that should be embedded into the replacement string.</param>
+        /// <param name="replacementString">The replacement string that contains the capture group markers (i.e. $1).</param>
+
+        public RegexStringTransformer(string captureGroupPattern, string replacementString)
+            : this(new (string, string)[] { (captureGroupPattern, replacementString) })
+        {
+        }
+
+        /// <summary>
+        /// Initializes the regex string transformer with the given capture group and replacement patterns.
+        /// </summary>
+        public RegexStringTransformer((string CaptureGroupPattern, string ReplacementString)[] rawPatterns)
+        {
+            _patterns = new (Regex Regex, string ReplacementString)[rawPatterns.Length];
+            for (int i = 0; i < rawPatterns.Length; i++)
+            {
+                _patterns[i] = (new Regex(rawPatterns[i].CaptureGroupPattern, RegexOptions.Compiled), rawPatterns[i].ReplacementString);
```

Comment:
Keep in mind you are taking in user-provided strings as patterns, that are going to work over user-provided inputs. This in Regex is usually an anti-pattern if not guarded well, as it could easily result in a ReDOS from a malicious user. At the very least, I would suggest passing in a matchTimeout to the Regex engines as a way to prevent at least some of this.